enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+
+#### Risk 3: Control plane node churn
+
+Automatically scaling up and down control planes could impact cluster
+reliability and availability, if performed too often.
+
+As such the additional logic must ensure that there is a grace period between
+changes to the machine size, during which no more scaling will be performed.
+
+### Drawbacks
+
+Some users might not want this feature and rather opt to scale control planes
+manually.
+
+## Design Details
+
+### Control Plane MachinSet operator configuration
+
+As automatic scaling will require adjustments to certain aspects of its logic,
+the implementation should implement a way for administrators of an OpenShift
+cluster to configure the specifics of how and when scaling will occur.
+
+This configuration will be performed using a new custom resource
+`ControlPlaneMachineSetAutoscaling`.
+
+The **required** configurations will include the following properties:
+- `machineConfiguration`: this YAML object copies the [
+  configuration](https://docs.openshift.com/container-platform/4.13/machine_management/control_plane_machine_management/cpmso-configuration.html)
+  from controlplane machinset operator to configure the cloud platform's
+  machinesetup with a machinetype and concrete values that must be adjusted. The
+  implementation should use multiple [discriminated
+  unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions).
+  One will specify the `machineType` used:
+  - `machineType`: currently always `machines_v1beta1_machine_openshift_io`, but
+  should be able to support [cluster API](https://cluster-api.sigs.k8s.io/) in
+  the future. 
+    - `machines_v1beta1_machine_openshift_io`: specifies options for the
+    machines API based provisioner.
+      - `machineConfigurations`: specifies the concrete specifications for the
+        machines.
+  
+  The `machineConfiguration` will use a discriminated union again to dispatch
+  for the cloud provider. The dispatch key will be `platform`. To start the
+  supported cloud providers must include:
+  - `Azure` and configuration for `vmSize`
+  - `GCP` and configuration for `machineType`
+  - `AWS` and configuration for `instanceType`
+  - `vSphere` and configuration for `memoryMiB`, `numCPUs` and `diskGiB`.
+  
+  Each element that denotes an instance configuration also has to include a
+  `weight` property to specify the order in which scaling should occur.
+
+The **optional** configurations will include the following properties:
+
+- `syncPeriod`: how often will the operator check resource usage to see if scaling
+  should be performed.
+- `scaleUp`: defines configuration for scaling up. This includes the following
+  sub elements:
+  - `stabilizationWindow`: duration to wait after a scale up before another
+    scaling operation can occur.
+  - `selectPolicy`: `next` to select the next bigger instance size.
+  - `triggerPolicy`: `all` or `any`. Determines if scaling occurs if one or
```

Comment:
Yes inside a `promql` query that should be possible with and and or.
I assume the most likely use case will be two metrics API triggers (one CPU, one memory) that should be 'ALL' or 'ANY' combined, or a complex prometheus metric to base the decision on.
If the metric needs to be combined, the user could even create a custom metrics based on a set of existing ones to make the query simpler.
So I would prefer not adding tags or anything to define if two triggers are ANDed or ORed together - but if that's needed maybe it could be added later.